src/app.rs

```diff
@@ -47,6 +47,138 @@ pub struct App {
+        // over the current session
+        self.receipt_set
+            .entry(event.uuid)
+            .or_insert_with(ReceiptQueues::new);
+
+        self.receipt_set
+            .get_mut(&event.uuid)
+            .unwrap()
+            .add(event.timestamp, event.receipt_type);
+    }
+
+    // Dictates whether receipts should be sent on the current tick
+    // Not used for now as
+    fn do_tick(&mut self) -> bool {
+        true
+    }
+
+    pub fn step(&mut self, signal_manager: &dyn SignalManager) -> bool {
+        if !self.do_tick() {
+            return false;
+        }
+        if self.receipt_set.is_empty() {
+            return false;
+        }
+        let uuid = *self.receipt_set.keys().next().unwrap();
+        let mut queue = self.receipt_set.remove(&uuid).unwrap();
+        if let Some((timestamps, receipt)) = queue.get_data() {
+            signal_manager.send_receipt(uuid, timestamps, receipt);
+            return true;
+        }
+        if !queue.is_empty() {
+            self.receipt_set.insert(uuid, queue);
+        }
+        false
+    }
+}
+
+/// This get built anywhere in the client and get passed to the App
+#[derive(Debug, Default, PartialEq, Eq)]
+pub struct ReceiptEvent {
+    uuid: Uuid,
+    /// Timestamp of the messages
+    timestamp: u64,
+    /// Type : Received, Read
+    receipt_type: Receipt,
+}
+
+impl ReceiptEvent {
+    pub fn new(uuid: Uuid, timestamp: u64, receipt_type: Receipt) -> Self {
+        Self {
+            uuid,
+            timestamp,
+            receipt_type,
+        }
+    }
+}
+
+#[derive(Debug, Default, PartialEq, Eq)]
+pub struct ReceiptQueues {
+    received_msg: HashSet<u64>,
+    read_msg: HashSet<u64>,
+}
+
+impl ReceiptQueues {
+    pub fn new() -> Self {
+        Self {
+            received_msg: HashSet::new(),
+            read_msg: HashSet::new(),
+        }
+    }
+
+    pub fn add_received(&mut self, timestamp: u64) {
+        if !self.received_msg.insert(timestamp) {
+            log::info!("Somehow got duplicate Received receipt @ {}", timestamp);
+        }
+    }
+
+    pub fn add_read(&mut self, timestamp: u64) {
+        // Ensures we do not send uselessly double the amount of receipts
+        // in the case a message is immediatly received and read.
+        self.received_msg.remove(&timestamp);
+        if !self.read_msg.insert(timestamp) {
+            log::info!("Somehow got duplicate Read receipt @ {}", timestamp);
+        }
+    }
+
+    pub fn add(&mut self, timestamp: u64, receipt: Receipt) {
+        match receipt {
+            Receipt::Received => self.add_received(timestamp),
+            Receipt::Read => self.add_read(timestamp),
+            _ => {}
+        }
+    }
+
+    pub fn _add_vec(&mut self, timestamps: Vec<u64>, receipt: Receipt) {
+        timestamps.into_iter().for_each(|t| self.add(t, receipt))
+    }
+
+    pub fn get_data(&mut self) -> Option<(Vec<u64>, Receipt)> {
```

Comment:
Just asking: shouln't it return received AND read msgs?

Reply:
It is possible, however that would require sending two different `ReceiptMessage`, as I understand it. So this is kinda the simple way of doing things

Reply:
Thank you for the explanation.